tapchannel/aux_funding.go

```diff
@@ -0,0 +1,933 @@
+	}
+
+	return nil
+}
+
+// newPendingChanID generates a new pending channel ID using a CSPRG.
+func newPendingChanID() (funding.PendingChanID, error) {
+	var id funding.PendingChanID
+	if _, err := io.ReadFull(crand.Reader, id[:]); err != nil {
+		return id, err
+	}
+
+	return id, nil
+}
+
+// pendingAssetFunding represents all the state needed to keep track of a
+// pending asset channel funding flow.
+type pendingAssetFunding struct {
+	peerPub btcec.PublicKey
+
+	pid funding.PendingChanID
+
+	initiator bool
+
+	// TODO(roasbeef): should be the new Either identifier?
+	assetID asset.ID
+
+	amt uint64
+
+	proofs []*proof.Proof
+
+	fundingRoot *mssmt.BranchNode
+
+	feeRate chainfee.SatPerVByte
+}
+
+// addProof adds a new proof to the set of proofs that'll be used to fund the
+// new channel.
+func (p *pendingAssetFunding) addProof(proof *proof.Proof) {
+	p.proofs = append(p.proofs, proof)
+}
+
+// assetRootFromInputs computes the asset root from the set of inputs provided.
+// This'll be used to identify the set of assets that'll be used as funding
+// inputs into the channel.
+func assetRootFromInputs(inputs []*proof.Proof) (*mssmt.BranchNode, error) {
+	ctxb := context.Background()
+
+	// Insert all the assets into a new SMT that'll commit to all the
+	// assets we plan to use as input to funding.
+	fundingTree := mssmt.NewCompactedTree(mssmt.NewDefaultStore())
+	for _, proof := range inputs {
+		assetInput := proof.Asset
+
+		assetKey := assetInput.AssetCommitmentKey()
+		assetLeaf, err := assetInput.Leaf()
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = fundingTree.Insert(ctxb, assetKey, assetLeaf)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// With all the items inserted, we can now compute the root that'll be
+	// used to identity this input set.
+	return fundingTree.Root(context.Background())
+}
+
+// bindFundingRoot binds the funding root to the pending channel ID. Once
+// bound, the pre funding process is complete, and as the responder, we're now
+// ready for the next phase of the funding flow.
+func (p *pendingAssetFunding) bindFundingRoot() error {
+	fundingRoot, err := assetRootFromInputs(p.proofs)
+	if err != nil {
+		return fmt.Errorf("error computing asset root: %v", err)
+	}
+
+	p.fundingRoot = fundingRoot
+
+	return nil
+}
+
+// assetOutputs returns the set of asset outputs that'll be used to fund the
+// new asset channel.
+func (p *pendingAssetFunding) assetOutputs() []*AssetOutput {
+	return fn.Map(p.proofs, func(p *proof.Proof) *AssetOutput {
+		return &AssetOutput{
+			AssetID: tlv.NewRecordT[tlv.TlvType0](
+				p.Asset.ID(),
+			),
+			Amount: tlv.NewPrimitiveRecord[tlv.TlvType1](
+				p.Asset.Amount,
+			),
+			Proof: tlv.NewRecordT[tlv.TlvType2](*p),
+		}
+	})
+}
+
+// newCommitBlob creates a new commitment blob that'll be stored in the channel
+// state for the specified party.
+func newCommitBlob(chanAssets assetOutputListRecord,
+	local bool) ([]byte, error) {
+
+	var commit Commitment
```

Comment:
We need to create proper proofs here, using the new commitment helper functions in my PR. Basically a single `Allocation` where all the funds go to one side. So we'll also need to know all funding keys and whether we're going to add (non-asset) commitment CPFP anchor outputs (which we'll need to create exclusion proofs for).

Reply:
Why are all the proofs needed at that level? My mental model is that here we just need to know which side owns which assets. 

The commitment key ring isn't yet known at this stage (that's where the aux leaf stuff comes into play).